File: vm_references/python-interpreter-common/src/rv32vm_pkg/protocol.py
"""Shared fixed little-endian persistent rv32vm protocol."""

from __future__ import annotations

import json
import struct

from .constants import (
    MAX_INPUT_LENGTH,
    MAX_INSTRUCTION_LIMIT,
    MAX_OUTPUT_LIMIT,
)
from .elf import ElfError
from .machine import LoadedProgram

MAGIC = b"RV32"
VERSION = 1
HEADER = struct.Struct("<4sBBHII")
RUN_HEADER = struct.Struct("<QII")
RUN_RESPONSE_HEADER = struct.Struct("<II")
MAX_PAYLOAD = 8 * 1024 * 1024

OP_READY = 0x80
OP_LOAD = 1
OP_RUN = 2
OP_RESET = 3
OP_UNLOAD = 4
OP_SHUTDOWN = 5

STATUS_OK = 0
STATUS_MALFORMED_FRAME = 1
STATUS_UNSUPPORTED_VERSION = 2
STATUS_UNKNOWN_OPCODE = 3
STATUS_INVALID_FLAGS = 4
STATUS_FRAME_TOO_LARGE = 5
STATUS_INVALID_PAYLOAD = 6
STATUS_INVALID_STATE = 7
STATUS_ELF_REJECTED = 8
STATUS_INTERNAL_ERROR = 9

_STATUS_MESSAGES = {
    STATUS_MALFORMED_FRAME: "malformed frame",
    STATUS_UNSUPPORTED_VERSION: "unsupported version",
    STATUS_UNKNOWN_OPCODE: "unknown opcode",
    STATUS_INVALID_FLAGS: "invalid flags",
    STATUS_FRAME_TOO_LARGE: "frame too large",
    STATUS_INVALID_PAYLOAD: "invalid payload",
    STATUS_INVALID_STATE: "invalid state",
    STATUS_ELF_REJECTED: "ELF rejected",
    STATUS_INTERNAL_ERROR: "internal error",
}


def result_bytes(result: dict) -> bytes:
    return json.dumps(
        result, ensure_ascii=True, allow_nan=False, separators=(",", ":")
    ).encode("utf-8")


def _write_response(
    stream, opcode: int, request_id: int, status: int, payload: bytes = b""
) -> None:
    stream.write(
        HEADER.pack(MAGIC, VERSION, opcode | 0x80, status, request_id, len(payload))
    )
    if payload:
        stream.write(payload)
    stream.flush()


def _read_exact(stream, length: int) -> bytes:
    chunks = bytearray()
    while len(chunks) < length:
        chunk = stream.read(length - len(chunks))
        if not chunk:
            break
        chunks.extend(chunk)
    return bytes(chunks)


def _error(stream, opcode: int, request_id: int, status: int, message: str) -> None:
    del message
    _write_response(
        stream, opcode, request_id, status, _STATUS_MESSAGES[status].encode("utf-8")
    )


def _run_response(
    program: LoadedProgram,
    input_data: bytes,
    instruction_limit: int,
    output_limit: int,
) -> bytes:
    machine = program.new_machine(input_data, output_limit)
    encoded_result = result_bytes(machine.run(instruction_limit))
    output = bytes(machine.output)
    return (
        RUN_RESPONSE_HEADER.pack(len(encoded_result), len(output))
        + encoded_result
        + output
    )


def serve(input_stream, output_stream) -> int:
    """Serve frames until a valid SHUTDOWN request or a fatal framing error."""
    program = None
    _write_response(output_stream, OP_READY, 0, STATUS_OK)

    while True:
        raw_header = _read_exact(input_stream, HEADER.size)
        if len(raw_header) != HEADER.size:
            if len(raw_header) >= 12:
                _error(
                    output_stream,
                    raw_header[5],
                    struct.unpack_from("<I", raw_header, 8)[0],
                    STATUS_MALFORMED_FRAME,
                    "malformed frame",
                )
            return 2
        magic, version, opcode, flags, request_id, payload_length = HEADER.unpack(
            raw_header
        )
        if magic != MAGIC:
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_MALFORMED_FRAME,
                "malformed frame",
            )
            return 2
        if payload_length > MAX_PAYLOAD:
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_FRAME_TOO_LARGE,
                "payload exceeds 8388608 bytes",
            )
            return 2
        payload = _read_exact(input_stream, payload_length)
        if len(payload) != payload_length:
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_MALFORMED_FRAME,
                "malformed frame",
            )
            return 2

        if version != VERSION:
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_UNSUPPORTED_VERSION,
                "unsupported protocol version",
            )
            continue
        if flags != 0:
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_INVALID_FLAGS,
                "request flags must be zero",
            )
            continue
        if opcode not in (OP_LOAD, OP_RUN, OP_RESET, OP_UNLOAD, OP_SHUTDOWN):
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_UNKNOWN_OPCODE,
                "unknown opcode",
            )
            continue

        try:
            if opcode == OP_LOAD:
                if not payload:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "invalid payload",
                    )
                    continue
                if program is not None:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_STATE,
                        "an image is already loaded",
                    )
                    continue
                try:
                    program = LoadedProgram(payload)
                except ElfError as error:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_ELF_REJECTED,
                        str(error),
                    )
                    continue
                _write_response(output_stream, opcode, request_id, STATUS_OK)
                continue

            if opcode == OP_RUN:
                if len(payload) < RUN_HEADER.size:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "RUN payload is shorter than 16 bytes",
                    )
                    continue
                instruction_limit, output_limit, input_length = RUN_HEADER.unpack_from(
                    payload
                )
                if len(payload) != RUN_HEADER.size + input_length:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "RUN input length does not match payload",
                    )
                    continue
                if instruction_limit > MAX_INSTRUCTION_LIMIT:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "instruction limit exceeds 100000000",
                    )
                    continue
                if output_limit > MAX_OUTPUT_LIMIT:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "output limit exceeds 1048576",
                    )
                    continue
                if input_length > MAX_INPUT_LENGTH:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "input length exceeds 4194304",
                    )
                    continue
                if program is None:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_STATE,
                        "no image is loaded",
                    )
                    continue
                response = _run_response(
                    program,
                    payload[RUN_HEADER.size :],
                    instruction_limit,
                    output_limit,
                )
                _write_response(output_stream, opcode, request_id, STATUS_OK, response)
                continue

            if opcode == OP_RESET:
                if payload:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "RESET payload must be empty",
                    )
                    continue
                if program is None:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_STATE,
                        "no image is loaded",
                    )
                    continue
                _write_response(output_stream, opcode, request_id, STATUS_OK)
                continue

            if opcode == OP_UNLOAD:
                if payload:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_PAYLOAD,
                        "UNLOAD payload must be empty",
                    )
                    continue
                if program is None:
                    _error(
                        output_stream,
                        opcode,
                        request_id,
                        STATUS_INVALID_STATE,
                        "no image is loaded",
                    )
                    continue
                program = None
                _write_response(output_stream, opcode, request_id, STATUS_OK)
                continue

            # SHUTDOWN is accepted in either loaded state.
            if payload:
                _error(
                    output_stream,
                    opcode,
                    request_id,
                    STATUS_INVALID_PAYLOAD,
                    "SHUTDOWN payload must be empty",
                )
                continue
            _write_response(output_stream, opcode, request_id, STATUS_OK)
            return 0
        except (MemoryError, OverflowError):
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_INTERNAL_ERROR,
                "internal server error",
            )
            return 2
        except Exception:  # noqa: BLE001
            # Do not expose host paths, exception classes, or other nondeterminism.
            _error(
                output_stream,
                opcode,
                request_id,
                STATUS_INTERNAL_ERROR,
                "internal server error",
            )
            return 2

File: vm_references/vm2-rust-interpreter/src/protocol.rs
//! Implements the persistent protocol defined by the `rv32vm` interface.

use std::io::{self, Read, Write};

use crate::machine::{LoadedProgram, MAX_INPUT_LENGTH, MAX_INSTRUCTION_LIMIT, MAX_OUTPUT_LIMIT};

const MAGIC: &[u8; 4] = b"RV32";
const VERSION: u8 = 1;
const HEADER_SIZE: usize = 16;
const RUN_HEADER_SIZE: usize = 16;
pub const MAX_PAYLOAD: usize = 8 * 1024 * 1024;

const OP_LOAD: u8 = 1;
const OP_RUN: u8 = 2;
const OP_RESET: u8 = 3;
const OP_UNLOAD: u8 = 4;
const OP_SHUTDOWN: u8 = 5;

const OK: u16 = 0;
const MALFORMED_FRAME: u16 = 1;
const UNSUPPORTED_VERSION: u16 = 2;
const UNKNOWN_OPCODE: u16 = 3;
const INVALID_FLAGS: u16 = 4;
const FRAME_TOO_LARGE: u16 = 5;
const INVALID_PAYLOAD: u16 = 6;
const INVALID_STATE: u16 = 7;
const ELF_REJECTED: u16 = 8;

fn u16_at(data: &[u8], offset: usize) -> u16 {
    u16::from_le_bytes([data[offset], data[offset + 1]])
}

fn u32_at(data: &[u8], offset: usize) -> u32 {
    u32::from_le_bytes([
        data[offset],
        data[offset + 1],
        data[offset + 2],
        data[offset + 3],
    ])
}

fn u64_at(data: &[u8], offset: usize) -> u64 {
    u64::from_le_bytes([
        data[offset],
        data[offset + 1],
        data[offset + 2],
        data[offset + 3],
        data[offset + 4],
        data[offset + 5],
        data[offset + 6],
        data[offset + 7],
    ])
}

fn status_message(status: u16) -> &'static [u8] {
    match status {
        MALFORMED_FRAME => b"malformed frame",
        UNSUPPORTED_VERSION => b"unsupported version",
        UNKNOWN_OPCODE => b"unknown opcode",
        INVALID_FLAGS => b"invalid flags",
        FRAME_TOO_LARGE => b"frame too large",
        INVALID_PAYLOAD => b"invalid payload",
        INVALID_STATE => b"invalid state",
        ELF_REJECTED => b"ELF rejected",
        _ => b"internal error",
    }
}

fn read_available<R: Read>(reader: &mut R, buffer: &mut [u8]) -> io::Result<usize> {
    let mut received = 0;
    while received < buffer.len() {
        match reader.read(&mut buffer[received..])? {
            0 => break,
            count => received += count,
        }
    }
    Ok(received)
}

fn write_response<W: Write>(
    writer: &mut W,
    opcode: u8,
    request_id: u32,
    status: u16,
    payload: &[u8],
) -> io::Result<()> {
    let mut header = [0; HEADER_SIZE];
    header[..4].copy_from_slice(MAGIC);
    header[4] = VERSION;
    header[5] = opcode | 0x80;
    header[6..8].copy_from_slice(&status.to_le_bytes());
    header[8..12].copy_from_slice(&request_id.to_le_bytes());
    header[12..16].copy_from_slice(&(payload.len() as u32).to_le_bytes());
    writer.write_all(&header)?;
    writer.write_all(payload)?;
    writer.flush()
}

fn write_error<W: Write>(
    writer: &mut W,
    opcode: u8,
    request_id: u32,
    status: u16,
) -> io::Result<()> {
    write_response(writer, opcode, request_id, status, status_message(status))
}

pub fn serve() -> Result<(), String> {
    let stdin = io::stdin();
    let stdout = io::stdout();
    serve_streams(&mut stdin.lock(), &mut stdout.lock())
}

fn serve_streams<R: Read, W: Write>(reader: &mut R, writer: &mut W) -> Result<(), String> {
    write_response(writer, 0, 0, OK, &[]).map_err(|error| error.to_string())?;
    let mut program = None;

    loop {
        let mut header = [0; HEADER_SIZE];
        let received = read_available(reader, &mut header).map_err(|error| error.to_string())?;
        if received != HEADER_SIZE {
            if received >= 12 {
                write_error(writer, header[5], u32_at(&header, 8), MALFORMED_FRAME)
                    .map_err(|error| error.to_string())?;
            }
            return Err("incomplete protocol header".into());
        }

        let opcode = header[5];
        let request_id = u32_at(&header, 8);
        if &header[..4] != MAGIC {
            write_error(writer, opcode, request_id, MALFORMED_FRAME)
                .map_err(|error| error.to_string())?;
            return Err("invalid protocol magic".into());
        }
        let payload_length = u32_at(&header, 12) as usize;
        if payload_length > MAX_PAYLOAD {
            write_error(writer, opcode, request_id, FRAME_TOO_LARGE)
                .map_err(|error| error.to_string())?;
            return Err("protocol frame is too large".into());
        }

        let mut payload = vec![0; payload_length];
        if read_available(reader, &mut payload).map_err(|error| error.to_string())?
            != payload_length
        {
            write_error(writer, opcode, request_id, MALFORMED_FRAME)
                .map_err(|error| error.to_string())?;
            return Err("incomplete protocol payload".into());
        }

        if header[4] != VERSION {
            write_error(writer, opcode, request_id, UNSUPPORTED_VERSION)
                .map_err(|error| error.to_string())?;
            continue;
        }
        if u16_at(&header, 6) != 0 {
            write_error(writer, opcode, request_id, INVALID_FLAGS)
                .map_err(|error| error.to_string())?;
            continue;
        }
        if !matches!(
            opcode,
            OP_LOAD | OP_RUN | OP_RESET | OP_UNLOAD | OP_SHUTDOWN
        ) {
            write_error(writer, opcode, request_id, UNKNOWN_OPCODE)
                .map_err(|error| error.to_string())?;
            continue;
        }

        match opcode {
            OP_LOAD => {
                if payload.is_empty() {
                    write_error(writer, opcode, request_id, INVALID_PAYLOAD)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                if program.is_some() {
                    write_error(writer, opcode, request_id, INVALID_STATE)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                match LoadedProgram::new(&payload) {
                    Ok(loaded) => program = Some(loaded),
                    Err(_) => {
                        write_error(writer, opcode, request_id, ELF_REJECTED)
                            .map_err(|error| error.to_string())?;
                        continue;
                    }
                }
                write_response(writer, opcode, request_id, OK, &[])
                    .map_err(|error| error.to_string())?;
            }
            OP_RUN => {
                if payload.len() < RUN_HEADER_SIZE {
                    write_error(writer, opcode, request_id, INVALID_PAYLOAD)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                let instruction_limit = u64_at(&payload, 0);
                let output_limit = u32_at(&payload, 8);
                let input_length = u32_at(&payload, 12) as usize;
                if payload.len() != RUN_HEADER_SIZE + input_length
                    || instruction_limit > MAX_INSTRUCTION_LIMIT
                    || output_limit > MAX_OUTPUT_LIMIT
                    || input_length > MAX_INPUT_LENGTH
                {
                    write_error(writer, opcode, request_id, INVALID_PAYLOAD)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                let Some(program) = &program else {
                    write_error(writer, opcode, request_id, INVALID_STATE)
                        .map_err(|error| error.to_string())?;
                    continue;
                };
                let mut machine = program.machine(&payload[RUN_HEADER_SIZE..], output_limit);
                let result = machine.run(instruction_limit).json();
                let mut response = Vec::with_capacity(8 + result.len() + machine.output.len());
                response.extend_from_slice(&(result.len() as u32).to_le_bytes());
                response.extend_from_slice(&(machine.output.len() as u32).to_le_bytes());
                response.extend_from_slice(result.as_bytes());
                response.extend_from_slice(&machine.output);
                write_response(writer, opcode, request_id, OK, &response)
                    .map_err(|error| error.to_string())?;
            }
            OP_RESET => {
                if !payload.is_empty() {
                    write_error(writer, opcode, request_id, INVALID_PAYLOAD)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                if program.is_none() {
                    write_error(writer, opcode, request_id, INVALID_STATE)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                write_response(writer, opcode, request_id, OK, &[])
                    .map_err(|error| error.to_string())?;
            }
            OP_UNLOAD => {
                if !payload.is_empty() {
                    write_error(writer, opcode, request_id, INVALID_PAYLOAD)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                if program.is_none() {
                    write_error(writer, opcode, request_id, INVALID_STATE)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                program = None;
                write_response(writer, opcode, request_id, OK, &[])
                    .map_err(|error| error.to_string())?;
            }
            OP_SHUTDOWN => {
                if !payload.is_empty() {
                    write_error(writer, opcode, request_id, INVALID_PAYLOAD)
                        .map_err(|error| error.to_string())?;
                    continue;
                }
                write_response(writer, opcode, request_id, OK, &[])
                    .map_err(|error| error.to_string())?;
                return Ok(());
            }
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use super::{HEADER_SIZE, MAGIC, OP_SHUTDOWN, VERSION, serve_streams};

    #[test]
    fn writes_ready_and_accepts_shutdown() {
        let mut request = [0; HEADER_SIZE];
        request[..4].copy_from_slice(MAGIC);
        request[4] = VERSION;
        request[5] = OP_SHUTDOWN;
        request[8..12].copy_from_slice(&7_u32.to_le_bytes());
        let mut output = Vec::new();

        serve_streams(&mut Cursor::new(request), &mut output).unwrap();

        assert_eq!(output.len(), HEADER_SIZE * 2);
        assert_eq!(&output[..4], MAGIC);
        assert_eq!(output[5], 0x80);
        assert_eq!(output[HEADER_SIZE + 5], OP_SHUTDOWN | 0x80);
        assert_eq!(u32::from_le_bytes(output[24..28].try_into().unwrap()), 7);
    }
}
